### DISCUSSION TITLE
How to structure integral constraints？

### DISCUSSION THREAD

**Discussion by kecbh** [lululxvi/deepxde#1299] [General]
The question is -u''=λu, λ is 1, and 0<x<pi, and u(0)=u(pi)=0.\int_0^{\pi} u(x)dx=\pi,I want to implement integral constraints, and there is currently no solution instability for integral constraint training.
Code as follow,
```
import deepxde as dde
import numpy as np
import matplotlib.pyplot as plt

geom = dde.geometry.Interval(0, np.pi)


def pde(x, y):
    dy_xx = dde.grad.hessian(y, x)
    return -dy_xx - y

def func(x):
    return np.sin(x) 

data = dde.data.PDE(geom, pde, [], num_domain=64, solution=func, num_test=400)

layer_size = [1] + [50] * 3 + [1]
activation = 'tanh'
initializer = 'Glorot uniform'
net = dde.nn.FNN(layer_size, activation, initializer)

def output_transform(x, y):
    return x *(np.pi-x) * y + x

net.apply_output_transform(output_transform)

model = dde.Model(data, net)
model.compile("adam", lr=1e-4, metrics=["l2 relative error"])

losshistory, train_state = model.train(iterations=5000)

dde.saveplot(losshistory, train_state, issave=True, isplot=True)
```

![Figure_2](https://user-images.githubusercontent.com/96463055/236626403-4e87d046-b299-48f7-82f1-6487a0fbf6e7.png)
I want to know how to add integral constraints to the code to guarantee the stability of the solution. I know you can use `model.compile,` but how do I do that? I don't know how deepxde is implemented.

**Comment by lululxvi** [comment]
Use `OperatorBC`. Check https://doi.org/10.1137/21M1397908 and code.

**Comment by yogibadepalli** [comment]
Dear @kecbh,

I'm currently working on implementing an integral constraint to normalize the output of a neural network, specifically using the following condition:

(\int_{0}^{\infty} |r^2| , dr = 1)

To achieve this, I need to incorporate the integral constraint into the training process (deepxde). I noticed from your earlier query that you were exploring a similar approach. I wanted to reach out and ask if you’ve found a solution or workaround for implementing this constraint effectively.

I don't think the functionality of operatorBC helps in adding integral constraint in the implementation.

Any insights or suggestions you can share would be greatly appreciated.

Best Regards,
Yogeswara.
